Add krakenuniq module

mjamy opened the pull request:
## PR checklist

Closes #102

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/taxprofiler/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/taxprofiler _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

jfy133 reviewed (changes requested):
Generally looks good so far:

- Missing `run_krakenuniq`, `krakenuniq_save_reads`, `krakenuniq_save_readclassifications`  default definition in `nextflow.config`
- Missing `run_krakenuniq`, `krakenuniq_save_reads`, `krakenuniq_save_readclassifications` in teh nextflow_schema.json (`nf-core schema build`, once you've updated the nextflow.config)
- Update `citations.md ` to include the KrakenUniq citation
- Update USAGE.md with the database information (I might be able to cover you on that one if you don't have a database yourself handy)
- Double check KrkaenUniq is listed in the README

mjamy commented:
Addressed comments:

> * Missing `run_krakenuniq`, `krakenuniq_save_reads`, `krakenuniq_save_readclassifications`  default definition in `nextflow.config`

Done
> * Missing `run_krakenuniq`, `krakenuniq_save_reads`, `krakenuniq_save_readclassifications` in teh nextflow_schema.json (`nf-core schema build`, once you've updated the nextflow.config)

Done
> * Update `citations.md ` to include the KrakenUniq citation

Done
> * Update USAGE.md with the database information (I might be able to cover you on that one if you don't have a database yourself handy)

I did this based on the krakenuniq databases available online. Would be good to double check this.
> * Double check KrkaenUniq is listed in the README

It is there.

jfy133 commented on docs/usage.md:
```diff
@@ -125,6 +126,20 @@ Expected (uncompressed) database files for each tool are as follows:
+  - `hash.k2d`
+  - `taxo.k2d`
+  - `database.idx`
+  - `database.kdb.counts`
+  - `database50mers.kmer_distrib`
+  - `database75mers.kmer_distrib`
+  - `database100mers.kmer_distrib`
+  - `database150mers.kmer_distrib`
+  - `database200mers.kmer_distrib`
+  - `database300mers.kmer_distrib`
+  - `seqid2taxid.map`
```
```suggestion
```

Because of: https://github.com/fbreitwieser/krakenuniq/blob/3efe086be6ea351624c1e946436a9b422864b8f8/scripts/build_db.sh#L351

jfy133 reviewed (approved):
Just need to fix linting (I'll do this for you @mjamy ) and you're FREEEEEEE

jfy133 commented:
Oh I guess I should test it as well :sweat_smile:

jfy133 commented:
@nf-core-bot fix linting

jfy133 commented:
Ok, hit some problems sadly:

- (pipeline) runs on a per-sample basis, rather than all files together
- (module) wrong preloaded flag for ram size (should be `--preload-size`)
- (module) doesn't correctly pick up paired end data (I didn't realise there was a paired-end) mode

Will need to think how to address this... maybe it won't be possible to run with reloaded mode :\

```
[91/f01647] Submitted process > NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:KRAKENUNIQ_PRELOADEDKRAKENUNIQ (2612_pe)
[0b/9ec147] Submitted process > NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:KRAKENUNIQ_PRELOADEDKRAKENUNIQ (2612_se)
[8b/683c5b] Submitted process > NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:KRAKENUNIQ_PRELOADEDKRAKENUNIQ (2613_pe)
[cb/40024c] Submitted process > NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:KRAKENUNIQ_PRELOADEDKRAKENUNIQ (2611_se)
[a8/2b5268] Submitted process > NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:KRAKENUNIQ_PRELOADEDKRAKENUNIQ (2612_pe)
[e4/31d6b3] Submitted process > NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:KRAKENUNIQ_PRELOADEDKRAKENUNIQ (ERR3201952_se)
Error executing process > 'NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:KRAKENUNIQ_PRELOADEDKRAKENUNIQ (2612_pe)'

Caused by:
  Process `NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:KRAKENUNIQ_PRELOADEDKRAKENUNIQ (2612_pe)` terminated with an error exit status (255)

Command executed:

  krakenuniq \
       \
      --db customdb_krakenuniq \
      --preload 16G \
      --threads 2
  
  for fastq in ERX5474932_ERR5766176_B_1.fastq.gz ERX5474932_ERR5766176_B_2.fastq.gz; do \
      PREFIX=$(echo $fastq)
      krakenuniq \
          --db customdb_krakenuniq \
          --threads 2 \
          --report-file "$PREFIX".krakenuniq.report.txt \
           \
           \
           \
           \
          --paired \
           \
          $fastq
  done
  
  
  
  cat <<-END_VERSIONS > versions.yml
  "NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:KRAKENUNIQ_PRELOADEDKRAKENUNIQ":
      krakenuniq: $(echo $(krakenuniq --version 2>&1) | sed 's/^.*KrakenUniq version //; s/ .*$//')
  END_VERSIONS

Command exit status:
  255

Command output:
  (empty)

Command error:
  /raven/ptmp/jfellowsy/nextflow/mjamy-taxprofiler/taxprofiler/testing/work/conda/env-b23453f9eff242ed726cc1cb1c5f6462/libexec/classify -d ./customdb_krakenuniq/database.kdb -i ./customdb_krakenuniq/database.idx -t 2 -M -a ./customdb_krakenuniq/taxDB -p 12
  Loading database(s)... 
   Database ./customdb_krakenuniq/database.kdb
  Loaded database with 29799 keys with k of 31 [val_len 4, key_len 8].
  
  complete.
  Reading taxonomy index from ./customdb_krakenuniq/taxDB. Done.
  classify: can't open 16G: No such file or directory
  krakenuniq: --paired requires exactly two filenames
```

jfy133 commented:
Currently borked because can't get the input looping through as pairs, but @Midnighter has a clever idea to make a file (one row for each paire) and while loop through that instead of the for loop

sofstam reviewed (approved):
Only information on `docs/OUTPUT.md` is missing but this is part of https://github.com/nf-core/taxprofiler/issues/18

jfy133 commented:
Thanks to @mjamy for the bulk of the work! And sorry for giving you one of the harder modules :see_no_evil:
